# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/SynchronousTask.h
# Commit: 58bf4c34d5c2
# Full Hash: 58bf4c34d5c2cef0c8034c901084c787985cc73b
# Author: Brad Werth <bwerth@mozilla.com>
# Date: 2022-11-29 21:47:49
# Description:
#   Bug 1801819: Make SynchronousTask double-check its completion flag to identify timeouts. r=sotaro
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D162877
# ==============================================================================

diff -r 8c350896c890 -r 58bf4c34d5c2 gfx/layers/ipc/SynchronousTask.h
--- a/gfx/layers/ipc/SynchronousTask.h	Tue Nov 29 05:06:46 2022 +0000
+++ b/gfx/layers/ipc/SynchronousTask.h	Tue Nov 29 05:15:07 2022 +0000
@@ -29,9 +29,16 @@
 
     // For finite timeouts, we only check once for completion, and otherwise
     // rely on the ReentrantMonitor to manage the interval. If the monitor
-    // returns too early, we'll never know.
+    // returns too early, we'll never know, but we can check if the mDone
+    // flag was set to true, indicating that the task finished successfully.
     if (!mDone) {
-      return mMonitor.Wait(aInterval);
+      // We ignore the return value from ReentrantMonitor::Wait, because it's
+      // always NS_OK, even in the case of timeout.
+      mMonitor.Wait(aInterval);
+
+      if (!mDone) {
+        return NS_ERROR_ABORT;
+      }
     }
 
     return NS_OK;
